### PR TITLE
Coin denylist v2 + config 

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 

- Deny list v2 using config 
- All APIs are behind public(protected) currently
- Types are public but will change before likely before hitting devnet

## Test plan 

- Added test scenario support and tests

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: sui-execution/latest/sui-move-natives/src/config.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        &MoveObjectType::from(field_setting_tag),
+    ) else {
+        return option_none(&value_ty);
+    };
+
+    let [_id, _name, setting]: [Value; 3] = unpack_struct(field)?;
+    let [data_opt]: [Value; 1] = unpack_struct(setting)?;
+    let data = match unpack_option(data_opt, &setting_data_value_ty)? {
+        None => {
+            // invariant violation?
```
Comment (tzakian) [inline sui-execution/latest/sui-move-natives/src/config.rs:138]: Need to think on it a bit more, but I think an invariant violation here makes sense.

**Review by tzakian — COMMENTED** [review]
Some initial comments. Overall things look pretty good though!

One thing worth making sure of is that there aren't any issues around loaded runtime objects and if that configs don't change the result of transaction dependencies and the like since I know we do some stuff around that higher up in e.g., the temporary store and the like. At least from this first pass it looks all good though, and I'll dive more deeply but something maybe to keep an eye out for.

File: crates/sui-framework/packages/sui-framework/sources/coin.move [hunk 2]
```diff
@@ -73,10 +77,9 @@ module sui::coin {
         total_supply: Supply<T>
     }
 
-    /// Capability allowing the bearer to freeze addresses, preventing those addresses from
-    /// interacting with the coin as an input to a transaction.
-    public struct DenyCap<phantom T> has key, store {
+    public struct DenyCapV2<phantom T> has key, store {
         id: UID,
+        allow_global_pause: bool,
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/coin.move:82]: It would be a new cap, and we would decide then what previous versions could update into.

File: crates/sui-framework/packages/sui-framework/sources/config.move [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        Value: copy + drop + store,
+    >(
+        config: & Config<WriteCap>,
+        name: Name,
+        ctx: &TxContext,
+    ): bool {
+        field::exists_with_type<_, Setting<Value>>(&config.id, name) && {
+            let epoch = ctx.epoch();
+            let sobj: &Setting<Value> = field::borrow(&config.id, name);
+            epoch == sobj.data.borrow().newer_value_epoch
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/config.move:109]: Was this comment for `new_for_epoch`? 
If so, I wanted to give something that indicated it would abort if already set, hence "new". Maybe `add`? Not sure

File: crates/sui-framework/packages/sui-framework/sources/coin.move [hunk 4]
```diff
@@ -293,56 +295,61 @@ module sui::coin {
...
-        if (type_name::is_primitive(&name)) return false;
+        let ty = type_name::get_with_original_ids<T>().into_string().into_bytes();
+        deny_list.v2_most_recent_contains(DENY_LIST_COIN_INDEX, ty, addr, ctx)
+    }
 
-        let `type` = type_name::into_string(name).into_bytes();
-        freezer.contains(DENY_LIST_COIN_INDEX, `type`, addr)
+    public(package) fun deny_list_v2_enable_global_pause<T>(
+        deny_list: &mut DenyList,
+        _deny_cap: &mut DenyCapV2<T>,
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/coin.move:329]: I did it to suppress the unused `&mut`, but I'll just annotate instead

File: sui-execution/latest/sui-move-natives/src/config.rs (hunk 1 above)
Comment (tnowacki) [reply sui-execution/latest/sui-move-natives/src/config.rs:138]: I am just very very scared of doing that here. Maybe just log it?

File: crates/sui-framework/packages/sui-framework/sources/config.move (hunk 3 above)
[the remark this answers is not in the document]
Comment (tzakian) [reply crates/sui-framework/packages/sui-framework/sources/config.move:109]: Not it was for this function -- I think the thing I got confused about when I first read this was whether it could ever be in an "unset" state in which case this function's name has a different meaning (is there a current value assigned to this `name` in the current epoch). But I realize now there must always be a value, so the name isn't as bad. 

That being said though, I still feel as though the name here is a bit hard to follow. IMO `set_in_current_epoch_for_type` or something like that would be a bit easier to follow.

File: crates/sui-framework/packages/sui-framework/sources/dynamic_object_field.move [hunk 5]
```diff
@@ -110,4 +95,105 @@ module sui::dynamic_object_field {
         let (_field, value_addr) = field::field_info<Wrapper<Name>>(object, key);
         option::some(value_addr.to_id())
     }
+
+    public(package) fun internal_add<Name: copy + drop + store, Value: key>(
```
Comment (cgswords) [inline crates/sui-framework/packages/sui-framework/sources/dynamic_object_field.move:99]: Nit: it took me a second to realize what these were doing different from the (more-public) versions. It may be worth writing a comment or something? (Or bikeshedding as `add_unstorable`...)

File: sui-execution/latest/sui-move-natives/src/config.rs (hunk 1 above)
Comment (tzakian) [reply sui-execution/latest/sui-move-natives/src/config.rs]: That's fine with me as well :) IMO this should either be a log or a panic, and I'm fine with either

File: crates/sui-framework/packages/sui-framework/sources/dynamic_object_field.move (hunk 5 above)
Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/dynamic_object_field.move:99]: The name here is assuming a new feature, namely `internal` type parameters

So it would be 
```
public fun internal_add<Name: copy + drop + store, internal Value: key>(
                                                   ^^^^^^^^ new thing here
```

Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/dynamic_object_field.move:99]: With that context, thoughts?

**Review by tzakian — APPROVED** [review]
LGTM!
